third_party/amd/lib/TritonAMDGPUTransforms/BlockPingpong.cpp

```diff
@@ -0,0 +1,286 @@
+          .failed() ||
+      genLocalSlice(builder, op.getB(), dotEncoding, 1, numSlices, sliceWidth)
+          .failed())
+    return failure();
+
+  // Clone dots four times to consume the slices
+  Operation *prevDot = op;
+  for (int i = 0; i < numSlices; i++) {
+    auto newOp = builder.clone(*op);
+    newOp->setOperand(0, loadSliceOps[0][i]->getResult(0));
+    newOp->setOperand(1, loadSliceOps[1][i]->getResult(0));
+    if (i > 0)
+      newOp->setOperand(2, prevDot->getResult(0));
+    prevDot = newOp;
+    dotSliceOps.push_back(newOp);
+  }
+  op->replaceAllUsesWith(prevDot);
+  op->erase();
+  for (auto loads : lLoadOps)
+    loads->erase();
+
+  builder.setInsertionPointAfter(gLoadOps[1]);
+  // Reorder operations into four mem/dot clusters
+
+  // mem1: global load A, local load A(1/4), local load B(1/4)
+  initOpInsertion(gLoadOps[1]);
+  attachSlicedLoadAB(/* slice */ 0);
+  attachClusterBarrier(builder, loc);
+
+  // dot (1/4)
+  attachOpWithPrio(builder, dotSliceOps[0], loc);
+  attachClusterBarrier(builder, loc);
+
+  // mem2: global load B, local load A(2/4), local load B(2/4)
+  attachOp(gLoadOps[1]);
+  attachSlicedLoadAB(/* slice */ 1);
+  attachClusterBarrier(builder, loc);
+
+  // dot (2/4)
+  attachOpWithPrio(builder, dotSliceOps[1], loc);
+  attachClusterBarrier(builder, loc);
+
+  // mem3: local load A(3/4, 4/4), local load B(3/4, 4/4)
+  attachSlicedLoadAB(/* slice */ 2);
+  attachSlicedLoadAB(/* slice */ 3);
+  attachClusterBarrier(builder, loc);
+
+  // dot (3/4)
+  attachOpWithPrio(builder, dotSliceOps[2], loc);
+  attachClusterBarrier(builder, loc);
+
+  // mem3: local store A and B
+  initOpInsertion(lStoreOps[1]);
+  attachClusterBarrier(builder, loc);
+
+  // dot (4/4)
+  attachOpWithPrio(builder, dotSliceOps[3], loc);
+  attachClusterBarrier(builder, loc);
+
+  return success();
+}
+
+void Pingponger::getDotPingponged() {
+  OpBuilder builder(forOp);
+  MLIRContext *ctx = forOp.getContext();
+  Location loc = forOp.getLoc();
+  auto f16_ty = builder.getF16Type();
+
+  forOp->walk([&](Operation *op) {
+    if (isa<triton::LoadOp>(op))
+      gLoadOps.push_back(op);
+    if (isa<ttg::LocalLoadOp>(op)) {
+      // Check if this is loading from pipeline
+      auto src = op->getOperand(0);
+      if (auto arg = mlir::dyn_cast<BlockArgument>(src))
+        if (auto tiedLoopInit = forOp.getTiedLoopInit(arg))
+          if (tiedLoopInit->get())
+            lLoadOps.push_back(op);
+    }
+    if (isa<ttg::LocalStoreOp>(op))
+      lStoreOps.push_back(op);
+    if (isa<triton::DotOp>(op))
+      dotOps.push_back(op);
+  });
+
+  if (gLoadOps.size() != 2 || lLoadOps.size() != 2 || dotOps.size() != 1)
+    return;
+
+  if (numWarps == 4) { // pingpong between warps from different blocks
+    transformOneWarp(builder, loc);
+  } else if (numWarps == 8) { // pingpong between warps from the same block
+    if (transformTwoWarp(builder, dotOps[0]->getLoc()).failed())
```

Comment:
I feel we can rename the current `transformTwoWarp` as `sliceDotForTwoWarp` and then create a new `transformTwoWarp` to wrap the whole `else` branch here.

Reply:
This is very good point, gives me an idea.
The transforms are not strictly limited to the number of warps but just how many pingpong clusters it schedules into. I'll rename them as `transformOnePPClusters` and `transformFourPPClusters`. So, in the later PR I'll add `transformTwoPPClusters` for the middle sized tile support.